src/main/scala/tech/sourced/gemini/ReportApp.scala

```diff
@@ -69,13 +74,31 @@ object ReportApp extends App {
         case `groupByMode` => ReportExpandedGroup(gemini.reportCassandraGroupBy(cassandra))
       }
 
-      val (connectedComponents, elsToBuckets) = gemini.findConnectedComponents(cassandra)
+      val (connectedComponents, elsToBuckets, elementIds) = gemini.findConnectedComponents(cassandra)
       saveConnectedComponents(connectedComponents, elsToBuckets, config.ccDirPath)
 
-      cassandra.close()
-      cluster.close()
+      val pythonCmd = s"python3 src/main/python/community-detector/report.py ${config.ccDirPath}"
```

Comment:
I suggest a very simple refactoring of this method, so it's logic is more transparent and maintainable in a long run:

```scala
// cassandra & gemini setup

val duplicateReport = ...
val similarityReport = ... // what is called `reportCommunities` now

print(duplicateReport)
printCommunities(similarityReport)
// cassandra teardown
```